File: cpp/examples/connection_pool_example.cpp
/**
 * @file connection_pool_example.cpp
 * @brief Example demonstrating UMICP Connection Pool
 */

#include "connection_pool.h"
#include "websocket_client.h"
#include <iostream>
#include <thread>
#include <chrono>
#include <vector>

using namespace umicp;

int main() {
    std::cout << "=== UMICP Connection Pool Example ===\n\n";

    // Create pool configuration
    std::cout << "1. Creating connection pool configuration...\n";
    PoolConfig config;
    config.address = "ws://localhost:8080";
    config.min_size = 2;
    config.max_size = 5;
    config.max_age_seconds = 300;        // 5 minutes
    config.idle_timeout_seconds = 120;   // 2 minutes
    config.acquire_timeout_ms = 5000;    // 5 seconds

    std::cout << "   Configuration:\n";
    std::cout << "   - Address: " << config.address << "\n";
    std::cout << "   - Min size: " << config.min_size << "\n";
    std::cout << "   - Max size: " << config.max_size << "\n";
    std::cout << "   - Max age: " << config.max_age_seconds << " seconds\n";
    std::cout << "   - Idle timeout: " << config.idle_timeout_seconds << " seconds\n";
    std::cout << "   - Acquire timeout: " << config.acquire_timeout_ms << " ms\n\n";

    // Create connection pool
    std::cout << "2. Creating connection pool...\n";
    ConnectionPool pool(config);
    std::cout << "   ✓ Pool created successfully\n";
    std::cout << "   Is closed: " << (pool.isClosed() ? "Yes" : "No") << "\n\n";

    // Show initial statistics
    std::cout << "3. Initial pool statistics:\n";
    auto stats = pool.getStats();
    std::cout << "   - Total connections: " << stats.total_connections << "\n";
    std::cout << "   - Available: " << stats.available_connections << "\n";
    std::cout << "   - In use: " << stats.in_use_connections << "\n";
    std::cout << "   - Total acquires: " << stats.total_acquires << "\n";
    std::cout << "   - Total releases: " << stats.total_releases << "\n";
    std::cout << "   - Total creates: " << stats.total_creates << "\n";
    std::cout << "   - Total closes: " << stats.total_closes << "\n";
    std::cout << "   - Failed acquires: " << stats.failed_acquires << "\n\n";

    // Note: The following operations would require a real WebSocket server
    std::cout << "4. Pool initialization:\n";
    std::cout << "   NOTE: Actual initialization requires a running WebSocket server\n";
    std::cout << "   This example demonstrates the API structure.\n\n";

    // Demonstrate pooled connection lifecycle
    std::cout << "5. Demonstrating pooled connection lifecycle:\n";
    WebSocketClientConfig ws_config;
    ws_config.url = "ws://localhost:8080";
    auto client = std::make_shared<WebSocketClient>(ws_config);
    auto conn = std::make_shared<PooledConnection>("ws://localhost:8080", client);

    std::cout << "   Connection created:\n";
    std::cout << "   - ID: " << conn->getId() << "\n";
    std::cout << "   - Address: " << conn->getAddress() << "\n";
    std::cout << "   - State: Available\n";
    std::cout << "   - Use count: " << conn->getUseCount() << "\n\n";

    std::cout << "   Acquiring connection...\n";
    conn->acquire();
    std::cout << "   - State: InUse\n";
    std::cout << "   - Use count: " << conn->getUseCount() << "\n\n";

    std::cout << "   Simulating work...\n";
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "   Releasing connection...\n";
    conn->release();
    std::cout << "   - State: Available\n";
    std::cout << "   - Use count: " << conn->getUseCount() << "\n\n";

    std::cout << "   Acquiring again...\n";
    conn->acquire();
    std::cout << "   - Use count: " << conn->getUseCount() << "\n\n";

    std::cout << "   Closing connection...\n";
    conn->close();
    std::cout << "   - State: Closed\n\n";

    // Demonstrate stale detection
    std::cout << "6. Testing stale connection detection:\n";
    WebSocketClientConfig test_config;
    test_config.url = "ws://localhost:8080";
    auto test_conn = std::make_shared<PooledConnection>("ws://localhost:8080",
                                                        std::make_shared<WebSocketClient>(test_config));

    std::cout << "   Fresh connection:\n";
    std::cout << "   - Is stale (10s timeout): " << (test_conn->isStale(10) ? "Yes" : "No") << "\n";
    std::cout << "   - Is idle (10s timeout): " << (test_conn->isIdle(10) ? "Yes" : "No") << "\n\n";

    std::cout << "   Waiting 2 seconds...\n";
    std::this_thread::sleep_for(std::chrono::seconds(2));

    std::cout << "   After 2 seconds:\n";
    std::cout << "   - Is stale (1s timeout): " << (test_conn->isStale(1) ? "Yes" : "No") << "\n";
    std::cout << "   - Is idle (1s timeout): " << (test_conn->isIdle(1) ? "Yes" : "No") << "\n\n";

    // Demonstrate pool cleanup
    std::cout << "7. Testing pool cleanup:\n";
    int removed = pool.cleanup();
    std::cout << "   ✓ Cleaned up " << removed << " connection(s)\n\n";

    // Pool configuration validation
    std::cout << "8. Testing configuration validation:\n";
    PoolConfig bad_config;
    bad_config.address = "ws://test:9000";
    bad_config.min_size = 20;  // Invalid: greater than max
    bad_config.max_size = 10;

    ConnectionPool test_pool(bad_config);
    std::cout << "   Original: min=" << bad_config.min_size << ", max=" << bad_config.max_size << "\n";
    std::cout << "   Adjusted: min=" << test_pool.getConfig().min_size
              << ", max=" << test_pool.getConfig().max_size << "\n";
    std::cout << "   ✓ Configuration automatically validated\n\n";

    // Connection states
    std::cout << "9. Connection state transitions:\n";
    WebSocketClientConfig state_config;
    state_config.url = "ws://localhost:8080";
    auto state_conn = std::make_shared<PooledConnection>("ws://localhost:8080",
                                                         std::make_shared<WebSocketClient>(state_config));

    auto print_state = [](PoolConnectionState state) {
        switch (state) {
            case PoolConnectionState::Available: return "Available";
            case PoolConnectionState::InUse: return "InUse";
            case PoolConnectionState::Validating: return "Validating";
            case PoolConnectionState::Closed: return "Closed";
            default: return "Unknown";
        }
    };

    std::cout << "   Initial state: " << print_state(state_conn->getState()) << "\n";

    state_conn->acquire();
    std::cout << "   After acquire: " << print_state(state_conn->getState()) << "\n";

    state_conn->release();
    std::cout << "   After release: " << print_state(state_conn->getState()) << "\n";

    state_conn->close();
    std::cout << "   After close: " << print_state(state_conn->getState()) << "\n\n";

    // Pool shutdown
    std::cout << "10. Shutting down pool...\n";
    pool.shutdown();
    std::cout << "    ✓ Pool shut down\n";
    std::cout << "    Is closed: " << (pool.isClosed() ? "Yes" : "No") << "\n\n";

    // Final statistics
    stats = pool.getStats();
    std::cout << "11. Final pool statistics:\n";
    std::cout << "    - Total connections: " << stats.total_connections << "\n";
    std::cout << "    - Available: " << stats.available_connections << "\n";
    std::cout << "    - In use: " << stats.in_use_connections << "\n";
    std::cout << "    - Total acquires: " << stats.total_acquires << "\n";
    std::cout << "    - Total releases: " << stats.total_releases << "\n";
    std::cout << "    - Total creates: " << stats.total_creates << "\n";
    std::cout << "    - Total closes: " << stats.total_closes << "\n";
    std::cout << "    - Failed acquires: " << stats.failed_acquires << "\n\n";

    std::cout << "=== Connection Pool Example Complete ===\n";
    std::cout << "\nNOTE: This example demonstrates the API structure.\n";
    std::cout << "For full functionality, connect to a running WebSocket server.\n";

    return 0;
}

File: bindings/java/umicp-examples/src/main/java/com/hivellm/umicp/examples/ConnectionPoolExample.java
package com.hivellm.umicp.examples;

import com.hivellm.umicp.transport.pool.*;

/**
 * Example demonstrating UMICP Connection Pool
 */
public class ConnectionPoolExample {
    public static void main(String[] args) throws Exception {
        System.out.println("=== UMICP Connection Pool Example ===\n");

        // Create pool configuration
        System.out.println("1. Creating connection pool configuration...");
        PoolConfig config = new PoolConfig()
                .setAddress("ws://localhost:8080")
                .setMinSize(2)
                .setMaxSize(5)
                .setMaxAgeSeconds(300)        // 5 minutes
                .setIdleTimeoutSeconds(120)   // 2 minutes
                .setAcquireTimeoutMs(5000);   // 5 seconds

        System.out.println("   Configuration:");
        System.out.println("   - Address: " + config.getAddress());
        System.out.println("   - Min size: " + config.getMinSize());
        System.out.println("   - Max size: " + config.getMaxSize());
        System.out.println("   - Max age: " + config.getMaxAgeSeconds() + " seconds");
        System.out.println("   - Idle timeout: " + config.getIdleTimeoutSeconds() + " seconds");
        System.out.println("   - Acquire timeout: " + config.getAcquireTimeoutMs() + " ms");
        System.out.println();

        // Create connection pool
        System.out.println("2. Creating connection pool...");
        try (ConnectionPool pool = new ConnectionPool(config)) {
            System.out.println("   ✓ Pool created successfully");
            System.out.println("   Is closed: " + pool.isClosed());
            System.out.println();

            // Show initial statistics
            System.out.println("3. Initial pool statistics:");
            PoolStats stats = pool.getStats();
            printStats(stats);

            // Pool initialization
            System.out.println("4. Pool initialization:");
            System.out.println("   NOTE: Actual initialization requires a running WebSocket server");
            System.out.println("   This example demonstrates the API structure.");
            System.out.println();

            // Demonstrate configuration validation
            System.out.println("5. Testing configuration validation:");
            PoolConfig badConfig = new PoolConfig("ws://test:9000");
            badConfig.setMinSize(20);
            badConfig.setMaxSize(10);

            System.out.println("   Original: min=" + badConfig.getMinSize() + ", max=" + badConfig.getMaxSize());

            badConfig.validate();

            System.out.println("   Adjusted: min=" + badConfig.getMinSize() + ", max=" + badConfig.getMaxSize());
            System.out.println("   ✓ Configuration automatically validated");
            System.out.println();

            // Demonstrate cleanup
            System.out.println("6. Testing pool cleanup:");
            int removed = pool.cleanup();
            System.out.println("   ✓ Cleaned up " + removed + " connection(s)");
            System.out.println();

            // Demonstrate cleanup task
            System.out.println("7. Background cleanup task:");
            pool.startCleanupTask(60);
            System.out.println("   ✓ Cleanup task started (60 second interval)");

            pool.stopCleanupTask();
            System.out.println("   ✓ Cleanup task stopped");
            System.out.println();

            // Multiple pool instances
            System.out.println("8. Testing multiple pool instances:");
            PoolConfig config1 = new PoolConfig("ws://localhost:8080");
            PoolConfig config2 = new PoolConfig("ws://localhost:8081");

            try (ConnectionPool pool1 = new ConnectionPool(config1);
                 ConnectionPool pool2 = new ConnectionPool(config2)) {

                System.out.println("   Pool 1: " + pool1.getConfig().getAddress());
                System.out.println("   Pool 2: " + pool2.getConfig().getAddress());
                System.out.println("   ✓ Multiple pools can coexist");
            }
            System.out.println();

            // Shutdown
            System.out.println("9. Shutting down pool...");
            pool.shutdown();
            System.out.println("   ✓ Pool shut down");
            System.out.println("   Is closed: " + pool.isClosed());
            System.out.println();

            // Final statistics
            System.out.println("10. Final pool statistics:");
            stats = pool.getStats();
            printStats(stats);

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }

        System.out.println("=== Connection Pool Example Complete ===");
        System.out.println("\nNOTE: This example demonstrates the API structure.");
        System.out.println("For full functionality, connect to a running WebSocket server.");
    }

    private static void printStats(PoolStats stats) {
        System.out.println("   - Total connections: " + stats.getTotalConnections());
        System.out.println("   - Available: " + stats.getAvailableConnections());
        System.out.println("   - In use: " + stats.getInUseConnections());
        System.out.println("   - Total acquires: " + stats.getTotalAcquires());
        System.out.println("   - Total releases: " + stats.getTotalReleases());
        System.out.println("   - Total creates: " + stats.getTotalCreates());
        System.out.println("   - Total closes: " + stats.getTotalCloses());
        System.out.println("   - Failed acquires: " + stats.getFailedAcquires());
        System.out.println();
    }
}

